### PR TITLE
Improve & update release process to reflect recent changes

### PR DESCRIPTION
**Pull request by ichard26**
### Description

- Formalise release cadence guidelines
- Overhaul release steps to be easier to follow and more thorough
- Reorder changelog template to something more sensible
- Update release automation docs to reflect recent improvements (notably the addition of in-repo mypyc wheel builds)

### Review notes

See https://ichard26-testblackdocs.readthedocs.io/en/overhaul-release-doc/contributing/release_process.html for a preview of this newly rewritten document.

@felix-hilden given you might be RMing the next release, it'd be great if you could check that this all makes sense to you! Thank you!

This also depends on #3223 as this PR assumes it's already been landed. Anyway I'll note my other potential points of discussion as comments.

### Checklist - did you ...

- [x] Add a CHANGELOG entry if necessary? -> this is internal only so n/a
- [x] Add / update tests if necessary? -> n/a
- [x] Add new / update outdated documentation? I heard this is a 100% documentation patch :)

### CODE REVIEW

File: docs/contributing/release_process.md
Comment (ichard26) [inline docs/contributing/release_process.md:9]: 1-2 months felt like the most reasonable given the current pace of Black's development, but I'm open to alternatives.

File: docs/contributing/release_process.md
Comment (ichard26) [inline docs/contributing/release_process.md:15]: Not entirely sure if we want to formalise this, but I feel like sometimes our development pace gets slow enough where releasing monthly stops being valuable.

File: docs/contributing/release_process.md
Comment (ichard26) [inline docs/contributing/release_process.md:37]: I tried to make this as easy as possible to follow while being thorough, but I might have made this too explicit and made the release process look scarier than it really is. I'd appreciate thoughts on it

File: docs/contributing/release_process.md
Comment (ichard26) [inline docs/contributing/release_process.md:96]: Thought it was worth making the distinction even clearer.

File: docs/contributing/release_process.md
Comment (ichard26) [inline docs/contributing/release_process.md:85]: In general I reordered the sections to what a regular user would find the most useful. Integrations got moved down because they aren't particularly relevant to most users. Packaging/parser changes on the other hand could probably either affect a lot of people and/or would be quite "newsworthy" (like supporting new syntax). 

Oh and docs really do not need the attention its fourth place gave it.

File: docs/contributing/release_process.md
Comment (felix-hilden) [inline docs/contributing/release_process.md:38]:
```suggestion
   - **_Black_ follows the [CalVer] versioning standard using the `YY.M.N` format**
```

File: docs/contributing/release_process.md
Comment (felix-hilden) [inline docs/contributing/release_process.md:205]:
```suggestion
[calver]: https://calver.org
```

File: docs/contributing/release_process.md
Comment (felix-hilden) [reply docs/contributing/release_process.md:85]: very nice 👍🏿

**Review by felix-hilden — APPROVED** [review]
Nicely done! I thought about reducing detail in some sections, but ya know what, those that don't need hand holding will just skip it! And it's easier on memory as well.

File: docs/contributing/release_process.md
Comment (JelleZijlstra) [inline docs/contributing/release_process.md:140]:
```suggestion
All of _Black_'s release automation uses [GitHub Actions]. All workflows are therefore
```

File: docs/contributing/release_process.md
Comment (JelleZijlstra) [inline docs/contributing/release_process.md:150]:
```suggestion
This is our main workhorse workflow. It builds an [sdist] and [wheels] to upload to PyPI
```
I pronounced it "ess-dist", so it should have "an"

File: docs/contributing/release_process.md
Comment (JelleZijlstra) [inline docs/contributing/release_process.md:156]:
```suggestion
This single job builds the sdist and pure Python wheel (i.e., a wheel that only contains
```

File: docs/contributing/release_process.md
Comment (JelleZijlstra) [inline docs/contributing/release_process.md:160]: What is the ellipsis for?

File: docs/contributing/release_process.md
Comment (JelleZijlstra) [inline docs/contributing/release_process.md:162]:
```suggestion
We use [mypyc] to compile _Black_ into a CPython C extension for significantly improved
```

File: docs/contributing/release_process.md
Comment (JelleZijlstra) [inline docs/contributing/release_process.md:171]: Same

File: docs/contributing/release_process.md
Comment (JelleZijlstra) [inline docs/contributing/release_process.md:189]:
```suggestion
The created binaries are stored on the associated GitHub Release for download
```

File: docs/contributing/release_process.md
Comment (JelleZijlstra) [inline docs/contributing/release_process.md:194]:
```suggestion
This workflow uses the QEMU powered `buildx` feature of Docker to upload an `arm64` and
```

**Review by JelleZijlstra — COMMENTED** [review]
Looks good, just some nits

File: docs/contributing/release_process.md
Comment (ichard26) [reply docs/contributing/release_process.md:160]: Supposed to be a placeholder for the names the matrix jobs will have in practice. 
image

I'm happy to remove them if you find them confusing.

File: docs/contributing/release_process.md
Comment (ichard26) [reply docs/contributing/release_process.md:171]: I was trying to add some separation between the following sentence and the previous section since the sdist / pure wheel job also uses twine to upload the distributions. 
image

Happy to take suggestions for better ways of making this clearer.

File: docs/contributing/release_process.md
Comment (ichard26) [reply docs/contributing/release_process.md]: Ah good catch!

File: docs/contributing/release_process.md
Comment (cooperlees) [inline docs/contributing/release_process.md:189]: Still makes me sad every time I read / get bit by it :(

File: docs/contributing/release_process.md
Comment (cooperlees) [inline docs/contributing/release_process.md:196]: Don't know if it's worth noting, but Ee from PSF also has access should I disappear.

**Review by cooperlees — APPROVED** [review]
Thanks! all sounds great to me.

I don't know if it fits here, but maybe a link to something that tells us how to debug cibuildwheel compiling failures might be handy too. But it's scope creep.

File: docs/contributing/release_process.md
Comment (ichard26) [reply docs/contributing/release_process.md:193]: I'm pretty Ee is our contact should we lose access to any of the platforms we use haha. They control the psf GitHub organisation, and are an admin of PyPI and the pyfound dockerhub organisation. It's probably not worth noting IMO.

**Comment by ichard26** [comment]
> I think it's fine to be verbose here - But we could make a quick release guide and link to more info? But yeah I'm open to either.

On second thought, I didn't like how verbose the proposed steps were so I tried to cut down on unnecessary fluff. While explicit/verbose steps may reduce mistakes, overly explicit steps can also lead to confusion so it's best to strike a good middle ground. 

I don't think a quick release guide is truly necessary. If you are really in a rush, presumably you won't read the steps you are already familiar with? And I don't think suggesting a bare minimum release process is worth it. Even if we're in a bind, you still have to file the release PR and draft (and publish) a GitHub Release (captured in the 10,000 foot view preamble). Sure, you could ignore the step telling you to wait for CI after merging the PR and also avoid copy-editing the changelog, but I don't believe it's really necessary to document that, no? 

I need to cut a release soon so unless someone objects or raises concerns I'll be merging this tomorrow east coast time. Final feedback would be appreciated.

**Comment by ichard26** [comment]
Oh and I forgot to say this in my previous comment. It's on my (ever-growing :upside_down_face:) to-do list to add some developer documentation on our usage of mypyc. Not sure what I will put on there, but linking to resources that help with debugging cibuildwheel failures sounds good. It's hard to give practical advice as builds can fail for a million different reasons, but I'll try to find/write something helpful :+1:

File: docs/contributing/release_process.md
Comment (ichard26) [reply docs/contributing/release_process.md]: Resolved in https://github.com/psf/black/pull/3242/commits/82b3765cdc39a97569783c635d84ace758f928c6.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**
